src/homomorphic_signature/mod.rs

```diff
@@ -0,0 +1,90 @@
+use core::ops::Add;
+
+use crate::Error;
+use ark_serialize::CanonicalSerialize;
+use ark_std::hash::Hash;
+use ark_std::rand::Rng;
+
+pub mod ncs1;
+pub trait HomomorphicSignatureScheme {
+    type Parameters: Clone + Send + Sync;
+    type PublicKey: CanonicalSerialize + Hash + Eq + Clone + Default + Send + Sync;
+    type SecretKey: CanonicalSerialize + Clone + Default;
+    type Signature: Clone + Default + Send + Sync + Add<Self::Signature, Output = Self::Signature>;
+    type Message: Clone + Default + Send + Sync + Add<Self::Message, Output = Self::Message>;
+    type Weight;
+
+    // Create single g2 element and n g1
+    fn setup<R: Rng>(rng: &mut R, n: usize) -> Result<Self::Parameters, Error>;
+
+    fn keygen<R: Rng>(
+        pp: &Self::Parameters,
+        rng: &mut R,
+    ) -> Result<(Self::PublicKey, Self::SecretKey), Error>;
+
+    fn sign(
+        pp: &Self::Parameters,
+        sk: &Self::SecretKey,
+        tag: &[u8],
+        index: &[u8],
+        message: &[Self::Message],
+        //rng: &mut R, TODO: Research if this makes sense to have per message randomness in HSS signing
```

Comment:
should we discuss this with Seve and Alexis?

Reply:
Only Seve I think, that's outside of my expertise.

Reply:
Seve is aware of this comment. It is actually removed in the upcoming PR.